### ISSUE TITLE
%feature("shadow") works on methods but not functions?

### ISSUE THREAD

**Issue by adl** [swig/swig#980]
I have been able to shadow Python wrappers for methods, but not functions.

For instance:

```
%module test

%feature("shadow") fun1 %{
def fun1(a, b):
  print("FUN1")
  return $action(a, b)
%}

%feature("shadow") fun2 %{
def fun2(a, b):
  print("FUN2")
  return $action(a, b)
%}

struct foo
{
  int fun1(int a, int b);
};

int fun2(int a, int b);
```

If I compile the above with `swig3.0 -c++ -python -py3 -O test.i` I get a `test.py` output that contains:

```python
class foo(object):
    ...
    def fun1(a, b):
      print("FUN1")
      return _test.foo_fun1(a, b)
...
def fun2(a: 'int', b: 'int') -> "int":
    return _test.fun2(a, b)
fun2 = _test.fun2
```

So the shadowing of `fun2` failed.  I've tried many variations about this (like `%feature("shadow") fun2(int,int) ...` or `%feature("shadow") ::fun2`) without success.

(I'm using Swig 3.0.10 from Debian unstable.)

**Comment by ojwb** [comment]
Indeed - it's only checked for methods.  I think the "shadow" in the name comes from "shadow class", which was what the proxy class in the target language used to be called.

This rather seems like a historical accident - I think a patch to address this would be welcome.

**Comment by ojwb** [comment]
Probably worth noting that you can use `%pythonprepend` and `%pythonappend` on both methods and functions, so this version of your example works:

```
%module test

%feature("shadow") fun1 %{
def fun1(a, b):
  print("FUN1")
  return $action(a, b)
%}

%pythonprepend fun2 %{
  print("FUN2")
%}

struct foo
{
  int fun1(int a, int b);
};

int fun2(int a, int b);
```

Not quite as flexible, but in practice you can probably achieve anything you could if `%feature("shadow")` worked.

**Comment by adl** [comment]
Yes.  The problem with `%pythonprepend` is that you need to now how the function arguments are named if you want to use them, and this naming has not been stable across Swig versions.

In fact, I'm actually using `%feature("shadow")` as a workaround because `%pythonprepend` failed on me in 2016.  My code, which simply inserts some exception-triggering precondition to the Python version of the function, still have this comment:

```c++
// Instead of %feature("shadow") we would like to use just
//   %pythonprepend spot::twa_graph::out %{ self.report_univ_dest(src) %}
// However Swig 3.0.2 (from Debian stable) names the argument "*args"
// while Swig 3.0.10 uses "src", and we want to be compatible with both.
%feature("shadow") spot::twa_graph::out %{
def out(self, src: 'unsigned int'):
    self.report_univ_dest(src)
    return $action(self, src)
%}
```
Maybe it's time that I forget about Swig 3 and clean this up :-)

**Comment by larryleguo** [comment]
I have the same issue. The following workaround works:
(1) ignore the function export.
(2) use %pythoncode to add the wrapper function you want.
